### PR TITLE
Framework: Lazy load optional components referenced in Layout

### PR DESCRIPTION
**Pull request by gziolo**
This PR is based on explorations done by @alexsanford in #14988. He extracted logged in layout to its own chunk. Seeing what gets bundles in this chunk I identified some code that can be lazy loaded:

screen shot 2017-06-20 at 12 36 47

This PR updates logged in `<Layout />` component to lazy load components that are loaded conditionally:
* NPS survey notice
* Guided tours
* Support user

and notifications box, which isn't required during initial render of `<Masterbar />`.

It reduces code to be loaded by user on production by **50 kBs** on initial page load and should speed up processing of JS code. In case of logged out users, those chunks won't be sent to the user. 

It would be nice to remove also `immutable.js` library, but it is handled in other PRs.

### Before
screen shot 2017-06-20 at 12 25 14

### After
screen shot 2017-06-20 at 10 47 03
screen shot 2017-06-20 at 10 52 43
screen shot 2017-06-20 at 10 49 47
screen shot 2017-06-20 at 10 54 10
screen shot 2017-06-20 at 10 55 53

### Testing
1. Open http://calypso.localhost:3000.
2. Click notifications icon and make sure it loads like before.
3. Type SU and try to log in as support user.
4. Type http://calypso.localhost:3000/post/?tour=editorBasicsTour to start guided tour (you still will have to pick a site).

I'm not sure how to test NPS survey notice. I'm also fine with leaving the latter in the build as it is relatively small.

### Review

* [ ] Code
* [ ] Product

### CODE REVIEW

**Comment by gziolo** [comment]
Good catch, we provide a default placeholder, which needs to be replaced with an empty component when loading in those cases.

**Comment by gziolo** [comment]
> I did see some odd visual artifacts though...

It should be fixed with 4dbe047.

File: client/components/async-load/index.jsx [hunk 1]
```diff
@@ -6,8 +6,13 @@ import { omit } from 'lodash';
...
 export default class AsyncLoad extends Component {
 	static propTypes = {
+		noPlaceholder: PropTypes.bool,
+		placeholder: PropTypes.node,
 		require: PropTypes.func.isRequired,
-		placeholder: PropTypes.node
+	};
+
+	static defaultProps = {
+		noPlaceholder: false,
```
Comment (dmsnell) [inline client/components/async-load/index.jsx]: couldn't we detect a `null` on `placeholder` and skip the boolean flag?

what should happen if we pass a valid placeholder _and_ `noPlaceholder: true`?

Comment (blowery) [inline client/components/async-load/index.jsx]: > couldn't we detect falsiness on placeholder and skip the boolean flag?

Right now, if you don't provide an explicit placeholder, we use a default placeholder. I _think_ it's ok to switch to requiring folks who want a placeholder to provide one, but it'll be a bigger change.

Comment (dmsnell) [inline client/components/async-load/index.jsx]: what about passing `null` then, since _not_ passing one is `undefined`?

Comment (mtias) [inline client/components/async-load/index.jsx]: We can just set default prop to be the placeholder component, so if you pass false/null it should work.

**Comment by mtias** [comment]
I like this approach much more.

File: client/components/async-load/index.jsx (hunk 1 above)
Comment (gziolo) [inline client/components/async-load/index.jsx]: I like that idea, I'll update tomorrow :)

**Comment by gziolo** [comment]
> the changes look good but I left a comment on the API for AsyncLoad, thinking we don't need to modify the ~~eternal~~ external part of it.

Updated as discussed above with d38d08a. You can pass `null` to skip displaying a default placeholder.

> we should note that this is making things harder to find in the code on account of no longer importing the modules in the file where they are used. before, for example, it was easy to search for `<Notifications` and get there. I wonder what we can do to mitigate this?

I was wondering the other day if we could mark component as async in it's definition. It would help to ensure that such component is always loaded asynchronously and therefore its code isn't duplicated in other chunks that want to load it synchronously. At the moment developer needs to make sure that a given component is always loaded in a certain way to help Webpack do all optimizations.

File: client/components/async-load/index.jsx [hunk 2]
```diff
@@ -6,8 +6,12 @@ import { omit } from 'lodash';
 
 export default class AsyncLoad extends Component {
 	static propTypes = {
+		placeholder: PropTypes.node,
 		require: PropTypes.func.isRequired,
-		placeholder: PropTypes.node
+	};
+
+	static defaultProps = {
+		placeholder: <div className="async-load__placeholder async-load" />,
```
Comment (gziolo) [inline client/components/async-load/index.jsx]: We should use `async-load__placeholder` only to make Eslint happy. We can open a follow up PR to cleanup that later.

**Comment by mtias** [comment]
> I was wondering the other day if we could mark component as async in it's definition. It would help to ensure that such component is always loaded asynchronously and therefore its code isn't duplicated in other chunks that want to load it synchronously.

I feel this is the choice of the person using the components/blocks. Some internal parts of a block can definitely be async loaded, as that is the choice of the block:

```jsx
<ACertainBlock>
    Bla
    <AsyncLoad somOtherBlock />
</ACertainBlock>

// Used by other people as:
<MyBlock>
    <ACertainBlock />
</MyBlock>
```

**Comment by gziolo** [comment]
> I feel this is the choice of the person using the components/blocks. Some internal parts of a block can definitely be async loaded, as that is the choice of the block:

I still don't know what is the best choice here, but definitely it's a off-topic here :)

File: client/layout/index.jsx [hunk 3]
```diff
@@ -39,16 +37,11 @@ import { isHappychatOpen } from 'state/ui/happychat/selectors';
 import SitePreview from 'blocks/site-preview';
 import { getCurrentLayoutFocus } from 'state/ui/layout-focus/selectors';
 import DocumentHead from 'components/data/document-head';
-import NpsSurveyNotice from 'layout/nps-survey-notice';
```
Comment (gziolo) [inline client/layout/index.jsx]: It's a general issue with how dynamic imports work :) Now it's also a part of language since they proposed 'import()'.

**Comment by mcsf** [comment]
Concerning Guided Tours, this PR looks good. It would be good, though, to test again after rebasing, so we can test with the latest tours. :-)

**Comment by mjuhasz** [comment]
> I'm not sure how to test NPS survey notice.

In environments where the `'nps-survey/dev-trigger'` config option is enabled (development and wpcalypso) you can run `npsSurvey();` in the console. That makes you eligible for seeing the NPS survey (which would otherwise be shown to select people only). If you are in a section that is whitelisted for the survey notice (see `SECTION_NAME_WHITELIST`) then the NPS survey notice will pop up after some delay.

For instance, if you're running Calypso locally, navigate to Blog Posts, and type `npsSurvey();` in the browser console, and after a short delay you will see the notice:

screen shot 2017-08-28 at 4 12 27 pm

**Comment by alisterscott** [comment]
It looks like this was addressed in  #18500 instead

**Pull request closed without merging**